### ISSUE TITLE
AprV2ModelCuda is Q4K-only by construction and wrong for Q4K — no configuration does useful GPU work

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#3923]
## Summary

`AprV2ModelCuda` — the `.apr` GPU class `apr chat` constructs — produces garbage
for Q4K models and **fails to initialise for everything else**. There is
currently no configuration in which it does useful GPU work.

Split out of #3922, whose release fix routes `chat` away from this class. This
ticket is the defect itself.

## Measured, one binary (`apr 0.69.1 (92858b7fd)`), one card, one session

| model | via `AprV2ModelCuda` (`apr chat --gpu`) | via the working path (`apr run --gpu`) |
|---|---|---|
| `qwen2.5-coder-1.5b-instruct-q4k.apr` | **garbage**, `ran=gpu fell_back=false` | correct on GPU, F2 PASSED |
| `…-q4k-v2.apr` | **garbage** (identical string) | — |
| `…-instruct-st.apr` (F32) | init fails → CPU | **correct on GPU**, F2 PASSED |
| `qwen2.5-coder-0.5b-instruct.apr` (bf16) | init fails → CPU | `rc=14` decline → CPU |

Reproducer:

```
printf 'What is the capital of France?\n/exit\n' \
  | apr chat qwen2.5-coder-1.5b-instruct-q4k.apr --json --max-tokens 16 --gpu

Assistant: ürnópez zombie.ERRópez zombieópez zombie zombie zombie zombie zombie
{"backend":{"requested":"default","ran":"gpu","fell_back":false}}   rc=0
```

`--json` is the non-interactive mode; a bare pipe without it hangs.

## The scope statement, which nobody had established

The init failure for the non-Q4K models is:

```
GH-279: Architecture weight validation failed:
  Invalid launch config: Quantized weight 'blk.0.attn_q.weight' not cached, will use CPU
```

**A quantized-weight validation failing on a model that has no quantized
weights** means the validator expects them. So the class is **Q4K-only by
construction** — and it is wrong for Q4K. Same mismatch shape as #3885
(`parse_apr_q4k_config` validating metadata only, so an all-f16 `.apr` was
accepted at spawn and failed per-request at prefill), caught here at init
instead.

A class in that state is fixed or removed, not left as a verb's default.

## Evidence about the nature of the defect

**The logits are numerically healthy and semantically wrong.** With
`APR_TRACE_LOGITS=1`, token 0 of the garbage run:

```
NaN: 0, Inf: 0, len: 151936
min: -13.6546, max: 13.9961, mean: 0.0023, var: 7.6293
kv_position: 15, kv_cache_len[0]: 15        (prompt was 15 tokens)
Top 5: [(60293, 13.996), (73319, 12.850), (130632, 12.656), (34918, 12.120), (15866, 11.947)]
```

That **eliminates a whole class of causes**: no overflow, no NaN propagation, no
uninitialised memory, no off-by-one in position tracking. The model is computing
something coherent from the wrong numbers — i.e. **misread weights**, not a blown
computation.

**Output is prompt-dependent, then collapses.** Three prompts gave
`"oooooooooooooooo usted…"`, `"negoci…"`, `"ópez…"` — different first tokens,
all degenerating to `zombie zombie zombie` within a few tokens. So input reaches
the model and generation degrades; it is not "the prompt never arrived".

**0.8 tok/s** (16 tokens in 20.9 s on a 4090 for a 1.5B Q4K). Whatever it is
doing, it is not a working GPU path, and the speed may share a cause with the
garbage.

## Candidates eliminated

* **`forward_single_cuda` ignoring its `position` argument is NOT the defect.**
  It is called (cache.rs:357) and does discard `_position`, but position is
  tracked internally via `self.kv_position`, which the trace confirms advances
  correctly (`kv_position: 15` for a 15-token prompt). Eliminated by measurement,
  not by reading.
* **Not the tokenizer.** `chat`/CPU loads the same sibling tokenizer and answers
  correctly.
* **Not a forbidden `*_colmajor` kernel.** None are reachable from the `.apr`
  GPU path; only doc comments and a guard asserting their absence.
* **Not q4k-read-as-F32 inflation.** The "24035 MB VRAM (F32 tensors)" log line
  is `vram_mb()` returning the *device's* total VRAM, misused by
  `print_apr_f32_perf_tip` — cosmetic, and worth fixing separately.

## Still open

Whether the fault is in the `.apr`→GPU upload (LAYOUT-001/002 at the format
boundary, `contracts/tensor-layout-v1.yaml`), the 890 MB F32 embedding cache the
working path does not build, or the forward itself.

## The control pair for whoever takes this

Same binary, same card, same file: **`apr run --gpu` correct, `apr chat --gpu`
garbage.** A must-GREEN and a must-RED that need no fixture.

Milestone 0.70.0 — #3922's routing fix removes this from the release path.
Related: #3922, #3885, #3921.

**Comment by noahgift** [comment]
milestone → 0.71.0 per docs/roadmaps/milestone-triage-2026-09-23.md (#4024), operator approval 2026-09-23 (relayed by aprender-cf). Reason: model/backend correctness or the certified matrix: Don't Leave Behind (#3994)
